posthog/async_migrations/migrations/0003_fill_person_distinct_id2.py

```diff
@@ -0,0 +1,90 @@
+from functools import cached_property
+
+from ee.clickhouse.client import sync_execute
+from posthog.async_migrations.definition import AsyncMigrationDefinition, AsyncMigrationOperation
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_DATABASE
+
+"""
+Migration summary:
+
+Schema change to migrate the data from the old person_distinct_id table
+to the new person_distinct_id2 table.
+
+The reason this is needed is for faster `person_distinct_id` queries as the
+old schema worked off of (distinct_id, person_id) pairs, making it expensive
+to for our analytics queries, which need to map from distinct_id -> latest person_id.
+
+The new schema works off of distinct_id columns, leveraging ReplacingMergeTrees
+with a version column we store in postgres.
+
+We migrate teams one-by-one to avoid running out of memory.
+
+The migration strategy:
+
+    1. write to both pdi and pdi2 any new updates (done prior to this migration)
+    2. insert all non-deleted (team_id, distinct_id, person_id) rows from pdi into pdi2 (this migration)
+    3. Once migration has run, we only read/write from/to pdi2.
+"""
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = "Set up person_distinct_id2 table, speeding up person-related queries."
+
+    depends_on = "0002_events_sample_by"
+
+    # After releasing this version we can remove code related to `person_distinct_id` table
+    posthog_max_version = "1.34.0"
+
+    def is_required(self):
+        rows = sync_execute(
+            """
+            SELECT comment
+            FROM system.columns
+            WHERE database = %(database)s AND table = 'person_distinct_id' AND name = 'distinct_id'
+        """,
+            {"database": CLICKHOUSE_DATABASE},
+        )
+
+        return len(rows) > 0 and rows[0][0] != "skip_0003_fill_person_distinct_id2"
```

Comment:
we can just `return True` if we mark the migration completed during new installs as I suggested above